test/rpm.jl/sb.jl

```diff
@@ -0,0 +1,100 @@
+using Turing
+using Turing.RandomMeasures
+using Test
+using Random
+
+# Data
+data = [-2,2,-1.5,1.5]
+
+# Base distribution
+mu_0 = mean(data)
+sigma_0 = 4
+sigma_1 = 0.5
+tau0 = 1/sigma_0^2
+tau1 = 1/sigma_1^2
+
+# DP parameters
+alpha = 0.25
+
+include("utils.jl")
+
+partitions = [[[1, 2, 3, 4]], [[1, 2, 3], [4]], [[1, 2, 4], [3]], [[1, 2], [3, 4]], [[1, 2], [3], [4]],
+[[1, 3, 4], [2]], [[1, 3], [2, 4]], [[1, 3], [2], [4]], [[1, 4], [2, 3]], [[1], [2, 3, 4]],
+[[1], [2, 3], [4]], [[1, 4], [2], [3]], [[1], [2, 4], [3]], [[1], [2], [3, 4]], [[1], [2], [3], [4]]]
+
+# truncated stick-breaking
+@model sbimm(y, rpm, trunc) = begin
+    # Base distribution.
+    H = Normal(mu_0, sigma_0)
+
+    # Latent assignments.
+    N = length(y)
+    z = tzeros(Int, N)
+    u = tzeros(Float64, N)
+
+    # Infinite collection of stick pieces and weights.
+    v = tzeros(Float64, trunc)
+    w = tzeros(Float64, trunc)
+    K = 0
+
+    # Cluster locations.
+    x = tzeros(Float64, trunc)
+
+    for i in 1:N
+
+        # Draw a slice ∈ [0,1].
+        u[i] ~ Beta(1, 1)
+
+        # Instantiate new cluster.
+        while (sum(w) < u[i]) && (K < trunc)
+            K += 1
+            v[K] ~ StickBreakingProcess(rpm)
+            x[K] ~ H
+            w[K] = v[K] * prod(1 .- v[1:(K-1)])
+        end
+
+        # Find truncation point
+        K_ = findfirst(u[i] .< cumsum(w))
```

Comment:
This seems like a non-typical slice sampler for DPs. Do you have a reference for this `slice` sampling representation?

Reply:
You are right. After looking at it again, it seems rather odd and is probably not quite correct. I can change it to the retrospective sampler by Papaspiliopoulos and Roberts which seems straight forward in Turing to me. Or do you have a preference for another one, e.g. Walker et al.?

Reply:
I don't have a preference for this; the method in Papaspiliopoulos and Roberts sounds good to me.

Reply:
Sounds good to me. I already started a BNP tutorial and will put the Papaspiliopoulos and Roberts code in there. With basic recursive stick breaking you mean a truncated implementation with fixed truncation point, right?

Reply:
I mean Alg 1 from the following paper: 

http://www.robots.ox.ac.uk/~twgr/assets/pdf/bloemreddy2017rpm.pdf

This version doesn't involve any truncation through the use of a random coin-flip based termination criterion.  I still need to read the original paper to have a better understanding of why this is equivalent to the standard stick-breaking, but my guess is that the expectation of the process converges to the standard stick-breaking process.

Reply:
I see. Yes, it looks like it does. I'll read the paper more carefully again as I forgot about the recursive coin-flipping.

Reply:
Algo 1 (coin-flipping based) is complicated to implement in Turing because of the variable name issue we have, i.e. the coin will not be resampled recursively because of: https://github.com/TuringLang/Turing.jl/blob/6c0f4b8e9cecc42d098b788f9dd7063da8ebcd3d/src/inference/Inference.jl#L114-L116  I'll keep the truncated stick-breaking test for now.

Reply:
I see, perhaps move this into a separate issue?

This should be fixed soon - see related discussion https://github.com/TuringLang/Turing.jl/pull/720#pullrequestreview-216110025.